### PR TITLE
[conv] Optimize performance of GWSS

### PR DESCRIPTION
**Pull request by atamazov**
- Optimizes GWSS for DYNAMIC_HYBRID mode
- Optimize out `IsApplicable()` calls when workspace is not necessary for a Solver
  - `MayNeedWorkspace()` is introduced in the [Solver API](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/866#issuecomment-822585406) for that.
- By-products
  - Refactor GetWorkspaceSize -> GetWorkspaceSizes for clarity

## Average acceleration of GWSS calls

- Immediate mode API (`miopenConvolutionForwardGetSolutionWorkspaceSize` etc): 15...115% 
- Find API (`miopenConvolutionForwardGetWorkSpaceSize` etc): 2...17%

However the overall effect on the total Aux Wall Time (that includes GWSS time) is relatively small and does not exceed 2%.

## Side effects

The total GWSS time for the use case described at #1297 is progressed from ~24 to ~7 ms.

:warning: This may camouflage the visibility of #1297 during GWSS calls. So engineers who work on #1297 may wish to revert this PR in their development trees.

### CODE REVIEW

**Comment by atamazov** [comment]
Please review.

File: src/include/miopen/solver.hpp [hunk 1]
```diff
@@ -2345,6 +2364,7 @@ struct GemmFwd1x1_0_2 : GemmFwdBase
     {
         return GetWorkspaceSize(ctx, ctx.conv_problem);
     }
+    bool MayNeedWorkspace() const { return true; }
```
Comment (averinevg) [inline src/include/miopen/solver.hpp:2367]: Wouldn't it be better to put MayNeedWorkspace() in the base class (GemmFwdBase)?

File: src/include/miopen/solver.hpp [hunk 2]
```diff
@@ -2463,6 +2487,7 @@ struct GemmBwd1x1_stride1 : GemmBwdBase
...
```
Comment (averinevg) [inline src/include/miopen/solver.hpp:2490]: Wouldn't it be better to put MayNeedWorkspace() in the base class (GemmBwdBase)?

File: src/include/miopen/solver.hpp [hunk 3]
```diff
@@ -2515,6 +2541,7 @@ struct GemmWrw1x1_stride1 : GemmWrwBase
...
```
Comment (averinevg) [inline src/include/miopen/solver.hpp:2544]: Wouldn't it be better to put MayNeedWorkspace() in the base class (GemmWrwBase)?

**Review by DrizztDoUrden — COMMENTED** [review]
LGTM

File: src/include/miopen/solver.hpp (hunk 1 above)
Comment (atamazov) [reply src/include/miopen/solver.hpp:2367]: That would create implicit dependence. `GemmFwdBase` does not provide a "base" implementation of  GetWorkspaceSize() and does not know if its successors need workspace.

**Comment by atamazov** [comment]
> LGTM

Hey, I need also a green mark ;)

**Review by DrizztDoUrden — APPROVED** [review]

**Comment by DrizztDoUrden** [comment]
My bad, misclicked.

**Comment by atamazov** [comment]
Results of testing added. @junliume Does the positive effect worth the slightly increased complexity of the code, what do you think?

**Review by junliume — APPROVED** [review]
As discussed, every perf optimization is good :)

**Pull request merged**